Mathlib/RingTheory/Spectrum/Prime/FreeLocus.lean

```diff
@@ -225,4 +227,125 @@ lemma isLocallyConstant_rankAtStalk [Module.FinitePresentation R M] [Module.Flat
 
+@[simp]
+lemma rankAtStalk_eq_zero_of_subsingleton [Subsingleton M] :
+    rankAtStalk (R := R) M = 0 := by
+  ext p
+  exact Module.finrank_zero_of_subsingleton
+
+lemma nontrivial_of_rankAtStalk_pos (h : 0 < rankAtStalk (R := R) M) :
+    Nontrivial M := by
+  by_contra! hn
+  have : Subsingleton M := not_nontrivial_iff_subsingleton.mp hn
+  simp [rankAtStalk, this] at h
+
+lemma rankAtStalk_eq_of_equiv {N : Type*} [AddCommGroup N] [Module R N] (e : M ≃ₗ[R] N) :
+    rankAtStalk (R := R) M = rankAtStalk N := by
+  ext p
+  exact IsLocalizedModule.mapEquiv p.asIdeal.primeCompl
+    (LocalizedModule.mkLinearMap p.asIdeal.primeCompl M)
+    (LocalizedModule.mkLinearMap p.asIdeal.primeCompl N) _ e |>.finrank_eq
+
+/-- If `M` is `R`-free, its rank at stalks is constant and agrees with the `R`-rank of `M`. -/
+@[simp]
+lemma rankAtStalk_eq_finrank_of_free [Module.Free R M] :
+    rankAtStalk (R := R) M = Module.finrank R M := by
+  ext p
+  simp [rankAtStalk, finrank_of_isLocalizedModule_of_free _ p.asIdeal.primeCompl
+    (LocalizedModule.mkLinearMap p.asIdeal.primeCompl M)]
+
+lemma rankAtStalk_self [Nontrivial R] : rankAtStalk (R := R) R = 1 := by
+  simp
+
+open LocalizedModule Localization
+
+/-- The rank of `Π i, M i` at a prime `p` is the sum of the ranks of `M i` at `p`. -/
+lemma rankAtStalk_pi {ι : Type*} [Finite ι] (M : ι → Type*)
+    [∀ i, AddCommGroup (M i)] [∀ i, Module R (M i)] [∀ i, Module.Flat R (M i)]
+    [∀ i, Module.FinitePresentation R (M i)] (p : PrimeSpectrum R) :
+    rankAtStalk (Π i, M i) p = ∑ᶠ i, rankAtStalk (M i) p := by
+  cases nonempty_fintype ι
+  let f : (Π i, M i) →ₗ[R] Π i, LocalizedModule p.asIdeal.primeCompl (M i) :=
+    .pi (fun i ↦ mkLinearMap p.asIdeal.primeCompl (M i) ∘ₗ LinearMap.proj i)
+  let e : LocalizedModule p.asIdeal.primeCompl (Π i, M i) ≃ₗ[Localization.AtPrime p.asIdeal]
+      Π i, LocalizedModule p.asIdeal.primeCompl (M i) :=
+    IsLocalizedModule.linearEquiv p.asIdeal.primeCompl
+      (mkLinearMap _ _) f |>.extendScalarsOfIsLocalization p.asIdeal.primeCompl _
+  have (i : ι) : Free (Localization.AtPrime p.asIdeal)
+      (LocalizedModule p.asIdeal.primeCompl (M i)) :=
+    free_of_flat_of_isLocalRing
+  simp_rw [rankAtStalk, e.finrank_eq, Module.finrank_pi_fintype, finsum_eq_sum_of_fintype]
+
+variable [Flat R M] [FinitePresentation R M]
+
+lemma rankAtStalk_eq_zero_iff_subsingleton :
+    rankAtStalk (R := R) M = 0 ↔ Subsingleton M := by
+  refine ⟨fun h ↦ ?_, fun _ ↦ rankAtStalk_eq_zero_of_subsingleton⟩
+  simp_rw [← support_eq_empty_iff (R := R), Set.eq_empty_iff_forall_not_mem, not_mem_support_iff]
+  intro p
+  have : Free (Localization.AtPrime p.asIdeal) (LocalizedModule p.asIdeal.primeCompl M) :=
+    free_of_flat_of_isLocalRing
+  apply subsingleton_of_rank_zero (R := Localization.AtPrime p.asIdeal)
+  have hr : finrank (Localization.AtPrime p.asIdeal) (LocalizedModule p.asIdeal.primeCompl M) = 0 :=
+    funext_iff.mp h p
+  simp [← finrank_eq_rank, hr]
```

Comment:
Can you move `rankAtStalk M p = 0 <-> not p in support R M` to a separate lemma?